libraries/app.rb

```diff
@@ -0,0 +1,150 @@
+#
+# Author:: Antoni S. Baranski (<antek.baranski@gmail.com>)
+# Copyright (C) 2016, ROBLOX, Inc.
+# License:: Apache License, Version 2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied.
+#
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Cookbook Name:: xcode
+# Library:: app
+#
+require 'chef/resource'
+require 'chef/provider'
+require 'chef/provider/ruby_block'
+
+module Xcode
+  module Resources
+    module XcodeApp
+      # The 'xcode_app' custom resource to install and uninstall Xcode components.
+      class Resource < Chef::Resource::LWRPBase
+        self.resource_name = :xcode_app
+        provides(:xcode_app)
+        actions(:install, :uninstall)
+        default_action(:install)
+
+        attribute(:version, kind_of: String, name_attribute: true)
+        attribute(:url, kind_of: String, required: true)
+        attribute(:checksum, kind_of: String, required: true)
+        attribute(:app, kind_of: String, required: true)
+        attribute(:multi_install, kind_of: [TrueClass, FalseClass], default: true)
+        attribute(:install_root, kind_of: String, default: '/Applications')
+        attribute(:force, kind_of: [TrueClass, FalseClass], default: false)
+      end
+
+      # The 'xcode_app' custom resource provider to install and uninstall Xcode components.
+      class Provider < Chef::Provider::LWRPBase
+        provides(:xcode_app)
+        use_inline_resources
+
+        action :install do
+          return if exist?
+          raise unless Chef.node.platform_family.eql?('mac_os_x')
+          install_app
+        end
+
+        action :uninstall do
+          directory install_dir do
+            recursive true
+            action :delete
+          end if exist?
+        end
+
+        def install_app
+          # If we force the install remove target folder if it exists.
+          directory install_dir do
+            recursive true
+            action :delete
+          end if new_resource.force && exist?
+
+          if new_resource.url.end_with?('dmg')
+            install_dmg
+          else
+            raise 'Unsupported package provided Xcode installer must be provided as a DMG'
+          end
+
+          post_install
+          accept_eula
```

Comment:
The EULA won't get accepted until the package install is complete even though EULA is not dependent on the packages. This is a nuanced edge case where: in a build farm setup, Xcode "looks" like it's installed (because it exists), a build kicks off, but would fail due to lack of EULA because package install takes a while (in some cases can be couple minutes). Fairly minor possibility of occurrence though (ex: disable build agent).